src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11SecretKeyFactory.java

```diff
@@ -286,22 +547,25 @@ protected SecretKey engineGenerateSecret(KeySpec keySpec)
         if (keySpec == null) {
             throw new InvalidKeySpecException("KeySpec must not be null");
         }
-        if (keySpec instanceof SecretKeySpec) {
+        if (keySpec instanceof SecretKeySpec secretKeySpec) {
             try {
-                Key key = convertKey(token, (SecretKey)keySpec, algorithm);
+                Key key = convertKey(token, secretKeySpec, algorithm);
                 return (SecretKey)key;
             } catch (InvalidKeyException e) {
                 throw new InvalidKeySpecException(e);
             }
+        } else if (keySpec instanceof PBEKeySpec pbeKeySpec &&
+                svcPbeKi != null) {
+            return (SecretKey) derivePBEKey(token, pbeKeySpec, svcPbeKi);
         } else if (algorithm.equalsIgnoreCase("DES")) {
-            if (keySpec instanceof DESKeySpec) {
-                byte[] keyBytes = ((DESKeySpec)keySpec).getKey();
+            if (keySpec instanceof DESKeySpec desKeySpec) {
+                byte[] keyBytes = desKeySpec.getKey();
```

Comment:
Would be nice to clear out "keyBytes" afterwards. Same goes for the other "keyBytes" in the same method.

Reply:
Good